docs/site/Middleware.md

```diff
@@ -0,0 +1,282 @@
+
+It's often desirable to reuse Express middleware in the sequence to handle API
+requests/responses without reinventing the wheel. We now add an
+`InvokeMiddleware` action as the first step in the default sequence. The action
+itself is an interceptor chain of `MiddlewareContext`. It uses the powerful
+[extension point/extension pattern](Extension-point-and-extensions.md) to
+discover registered Express middleware and invoke them as a chain.
+
+#### Default sequence
+
+The generated `src/sequence.ts` now has a new `middleware` property to
+`DefaultSequence` to allow injection of the middleware chain while keeping
+backward compatibility of the constructor.
+
+{% include code-caption.html content="src/sequence.ts" %}
+
+```ts
+export class DefaultSequence implements SequenceHandler {
+  /**
+   * Optional middleware chain
+   * Invokes registered middleware (injected via SequenceActions.MIDDLEWARE).
+   */
+  @inject(SequenceActions.MIDDLEWARE, {optional: true})
+  protected invokeMiddleware: InvokeMiddleware = () => false;
+
+  // ...
+  async handle(context: RequestContext): Promise<void> {
+    try {
+      const {request, response} = context;
+      const finished = await this.invokeMiddleware(context);
+      if (finished) return;
+      const route = this.findRoute(request);
+      const args = await this.parseParams(request, route);
+      const result = await this.invoke(route, args);
+
+      debug('%s result -', route.describe(), result);
+      this.send(response, result);
+    } catch (error) {
+      this.reject(context, error);
+    }
+  }
+}
+```
+
+#### Extend sequence with more than one middleware actions
+
+Sometimes we want to add middleware to the sequence in between other actions,
+for example, do some post-processing before the result is written to the HTTP
+response. This can be achieved by overriding the sequence implementation to call
+`invokeMiddleware` as needed.
+
+{% include code-caption.html content="src/sequence.ts" %}
+
+```ts
+export class SequenceWithMiddleware extends DefaultSequence {
+  async handle(context: RequestContext): Promise<void> {
+    try {
+      const {request, response} = context;
+      // The default middleware chain
+      await this.invokeMiddleware(context);
+      const route = this.findRoute(request);
+      const args = await this.parseParams(request, route);
+      const result = await this.invoke(route, args);
+
+      // The second middleware chain for post-invocation processing
+      context.bind('invocation.result').to(result);
+      // POST_INVOCATION_MIDDLEWARE is the name of the extension point
+      // for post-invocation middleware
+      await this.invokeMiddleware(context, {
+        extensionPoint: POST_INVOCATION_MIDDLEWARE,
+      });
+      this.send(response, result);
+    } catch (error) {
+      this.reject(context, error);
+    }
+  }
+}
+```
+
+It's also possible to inject multiple instances of `InvokeMiddleware`, each of
+which has its own extension point name.
```

Comment:
What is the `different` extension point name being used below? It is not obvious.